### PR TITLE
Feature/restaurants page

### PR DESCRIPTION
**Pull request by OrenMoveo**
# Restaurants Page Feature

## This PR Includes
- Responsive UI design for desktop and mobile devices.
- Fliter button functionality on mobile and desktop.
- Page routing for the homepage and restaurants page.
- Rating dropdown popover is complete.

## TODO
- Fix onBlur for all dropdowns popovers. 
- Add pricing range and distance popover UI.

## Examples

https://github.com/OrenMoveo/Epicure/assets/156906187/6efb0067-7863-414d-a172-f3ef194c42ab

*Mobile*

https://github.com/OrenMoveo/Epicure/assets/156906187/0e582085-a135-42c2-9825-1bd8c4070af8

*Desktop*

### CODE REVIEW

File: src/types/types.ts [hunk 1]
```diff
@@ -6,7 +7,9 @@ export interface Restaurant {
   location: string; // not sure
   rating: number;
   priceRange: number[];
+  mostPopular: boolean;
   keyId: string;
+  new: boolean;
```
Comment (nitzanpap) [inline src/types/types.ts]: `new` is a bad word, many times it is a key word of the language, and will confuse other developers, or might even break your code

File: src/pages/Root/Root.module.scss
Comment (nitzanpap) [inline src/pages/Root/Root.module.scss:1]: Delete empty files, we went over this.

File: src/components/RestaurantCard/RestaurantCard.tsx [hunk 2]
```diff
@@ -1,35 +1,42 @@
...
   const ratingImageSrcString: string = `/src/assets/images/ratings/rating${restaurant.rating}.svg`;
 
   return (
-    <div className="card-container">
-      <div className="card-image-container">
+    <button
+      className={`${styles.restaurantCardContainer} ${
+        className ? styles[className] : ""
+      }`}
+    >
```
Comment (nitzanpap) [inline src/components/RestaurantCard/RestaurantCard.tsx]: Remind me to go over this with you. There's a better way.

File: src/pages/RestaurantsPage/RestaurantsPage.tsx [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  const [rating, setRating] = useState(3);
+  const [priceRange, setPriceRange] = useState([0, 330]);
+  const [restaurants, setRestaurants] = useState<Restaurant[]>(restaurantsData);
+  const [activeFilterButton, setActiveFilterButton] = useState<number>(0);
+
+  const ALL_INDEX = 0;
+  const NEW_INDEX = 1;
+  const MOST_POPULAR_INDEX = 2;
+  const OPEN_NOW_INDEX = 3;
+  const MAP_VIEW_INDEX = 4;
```
Comment (nitzanpap) [inline src/pages/RestaurantsPage/RestaurantsPage.tsx]: Use `enum`s instead, or `const {} as const` objects.

File: src/pages/RestaurantsPage/RestaurantsPage.tsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  const restaurantsData: Restaurant[] = data.restaurants;
+  const isTablet = useIsTablet();
+  const isMobile = useIsMobile();
+  const isMobileOrTable = isMobile || isTablet;
+  const buttonRefs: React.MutableRefObject<HTMLButtonElement | null>[] = [
+    useRef(null),
+    useRef(null),
+    useRef(null),
+    useRef(null),
+  ];
```
Comment (nitzanpap) [inline src/pages/RestaurantsPage/RestaurantsPage.tsx]: I highly doubt you really need to use `refs` in here.

File: src/components/RestaurantCard/RestaurantCard.tsx [hunk 5]
```diff
@@ -1,35 +1,42 @@
...
 interface RestaurantCardProps {
   restaurant: Restaurant;
+  className?: string;
 }
 
-const RestaurantCard = ({ restaurant }: RestaurantCardProps) => {
+const RestaurantCard = ({ restaurant, className }: RestaurantCardProps) => {
   const screenWidth = useGetScreenWidth();
   const isTablet = screenWidth < UIConstants.sizes.tabletWidth;
   const ratingImageSrcString: string = `/src/assets/images/ratings/rating${restaurant.rating}.svg`;
```
Comment (nitzanpap) [inline src/components/RestaurantCard/RestaurantCard.tsx]: This is a bad way to import something. This absolute path creates an opening for serious bugs. Can you see why?

File: src/components/GoToAllRestaurantsButton/GoToAllRestaurantsButton.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import styles from "./GoToAllRestaurantsButton.module.scss";
+import goToIcon from "../../assets/images/goToIcon.svg";
+import { Link } from "react-router-dom";
+
+const GoToAllRestaurantsButton = () => {
+  return (
+    <Link to="/restaurants" className={styles.linkContainer}>
```
Comment (nitzanpap) [inline src/components/GoToAllRestaurantsButton/GoToAllRestaurantsButton.tsx]: What do you think could be the problem with this link?

File: src/components/Dropdown/Dropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+              <Rating
+                rating={4}
+                isChecked={isChecked}
+                setIsChecked={setIsChecked}
+              />
+              <Rating
+                rating={5}
+                isChecked={isChecked}
+                setIsChecked={setIsChecked}
+              />
```
Comment (nitzanpap) [inline src/components/Dropdown/Dropdown.tsx]: Why not use a loop? Or better yet, why not create a component that does this?

File: src/components/Dropdown/Dropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    return booleanArray.some((value) => value === true);
+  };
+
+  useEffect(() => {
+    if (isOpen) {
+      ratingPopoverRef.current?.focus();
+    }
+  }, [isOpen]);
+
+  function handleClearAll(): void {
```
Comment (nitzanpap) [inline src/components/Dropdown/Dropdown.tsx]: Use ES6 syntax. Be consistent.

File: src/components/Dropdown/Dropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    false,
+    false,
+    false,
+    false,
+    false,
+  ]);
+  const ratingPopoverRef = useRef<HTMLDivElement | null>(null);
+
+  const hasAtLeastOneTrue = (booleanArray: boolean[]) => {
+    return booleanArray.some((value) => value === true);
```
Comment (nitzanpap) [inline src/components/Dropdown/Dropdown.tsx]: You can just write `booleanArray.some((value) => value)` as the `some()` method will automatically check for the boolean value of the returned value.

File: src/App.tsx [hunk 10]
```diff
@@ -1,8 +1,32 @@
...
+      {
+        path: "/restaurants",
+        element: <RestaurantsPage />,
+      },
+      {
+        path: "/chefs",
+      },
+    ],
+  },
+]);
```
Comment (nitzanpap) [inline src/App.tsx]: What do you think could be better?

**Review by nitzanpap — CHANGES_REQUESTED** [review]
Let's go over these quickly.

File: src/types/types.ts (hunk 1 above)
Comment (OrenMoveo) [reply src/types/types.ts]: @nitzanpap  done

File: src/pages/Root/Root.module.scss
Comment (OrenMoveo) [reply src/pages/Root/Root.module.scss:1]: @nitzanpap  done

File: src/pages/RestaurantsPage/RestaurantsPage.tsx (hunk 3 above)
Comment (OrenMoveo) [reply src/pages/RestaurantsPage/RestaurantsPage.tsx]: @nitzanpap  done

File: src/pages/RestaurantsPage/RestaurantsPage.tsx (hunk 4 above)
Comment (OrenMoveo) [reply src/pages/RestaurantsPage/RestaurantsPage.tsx]: @nitzanpap done

File: src/components/Dropdown/Dropdown.tsx (hunk 8 above)
Comment (OrenMoveo) [reply src/components/Dropdown/Dropdown.tsx]: @nitzanpap  done

File: src/components/RestaurantCard/RestaurantCard.tsx (hunk 5 above)
Comment (OrenMoveo) [reply src/components/RestaurantCard/RestaurantCard.tsx]: @nitzanpap  done

File: src/components/GoToAllRestaurantsButton/GoToAllRestaurantsButton.tsx (hunk 6 above)
Comment (OrenMoveo) [reply src/components/GoToAllRestaurantsButton/GoToAllRestaurantsButton.tsx]: @nitzanpap  done

File: src/components/Dropdown/Dropdown.tsx (hunk 7 above)
Comment (OrenMoveo) [reply src/components/Dropdown/Dropdown.tsx]: @nitzanpap  done

File: src/components/Dropdown/Dropdown.tsx (hunk 9 above)
Comment (OrenMoveo) [reply src/components/Dropdown/Dropdown.tsx]: @nitzanpap  done

File: src/App.tsx (hunk 10 above)
Comment (OrenMoveo) [reply src/App.tsx]: @nitzanpap done

**Pull request merged**